File: day_11/merge_sort.py
"""
Merge Sort-You are given an array of integers. 
Write a function that will take this array as input and return the sorted array 
using Merge sort.
"""

def merge_sort(arr):
  if(len(arr) > 1):
    left = 0
    right = len(arr)
    mid = (left + right) // 2

    left_arr = arr[left: mid]
    right_arr = arr[mid: right]

    merge_sort(left_arr)
    merge_sort(right_arr)

    i = 0
    j = 0
    k = 0

    while (i < len(left_arr) and j < len(right_arr)):
      if (left_arr[i] <= right_arr[j]):
        arr[k] = left_arr[i]
        i += 1
      else:
        arr[k] = right_arr[j]
        j += 1
      k += 1

    while (i < len(left_arr)):
      arr[k] = left_arr[i]
      i += 1
      k += 1

    while (j < len(right_arr)):
      arr[k] = right_arr[j]
      j += 1
      k += 1

  return arr

a = [4, 56, 34, 10, 77, 51, 93, 30, 5, 52]
b = [72, 34, 88, 14, 32, 12, 34, 77, 56, 83]
print(merge_sort(a))
print(merge_sort(b))

File: day_11/mergeSort.js
/**
 * Merge Sort-You are given an array of integers. 
 * Write a function that will take this array as input and return the sorted array 
 * using Merge sort.
 */

function mergeSort(arr) {
  if (arr.length > 1){
    const left = 0;
    const right = arr.length;

    const mid = Math.floor((left + right) / 2);

    const leftArr = arr.slice(left, mid);
    const rightArr = arr.slice(mid, right);

    // recursion
    mergeSort(leftArr);
    mergeSort(rightArr);

    let i = 0;
    let j = 0;
    let k = 0;

    while(i < leftArr.length && j < rightArr.length) {
      if(leftArr[i] >= rightArr[j]){
        arr[k++] = rightArr[j++];
      } else {
        arr[k++] = leftArr[i++];
      }
    }

    while (i < leftArr.length) {
      arr[k++] = leftArr[i++];
    }

    while (j < rightArr.length) {
      arr[k++] = rightArr[j++];
    }
  }
  return arr;
}

const a = [4, 56, 34, 10, 77, 51, 93, 30, 5, 52];
const b = [72, 34, 88, 14, 32, 12, 34, 77, 56, 83];
console.log(mergeSort(a));
console.log(mergeSort(b));